### ISSUE TITLE
do not recommend source() within rmarkdown

### ISSUE THREAD

**Issue by aaronpeikert** [cjvanlissa/worcs#75]
A user approached @brandmaier with a technical problem (regarding caching) with our reproducibility workflow that we tracked down to the use of `source()` within RMarkdown.

https://github.com/cjvanlissa/worcs/blob/d5c21b2514e9b896b969c8ea673f1a156b138b0a/paper/Manuscript.Rmd#L225

I personally much favour using `knitr::read_chunk()` because it is more transparent and reproducible.

If we get an opportunity to change it, I would love to see it.

**Comment by cjvanlissa** [comment]
OK, lets do it!

**Comment by cjvanlissa** [comment]
I'm a bit apprehensive to do this, because the chunk files have again a slightly different syntax, which I think is confusing. I think the `source()` function more closely matches what novice users do in practice: Just writing a script file interactively. Can you clarify why using `source()` is bad practice so I can make an informed decision?

**Comment by aaronpeikert** [comment]
Sorry for replying so late.
This particular case was about the problem that using `source()` completely breaks knitr's caching.
However, there are other much harder to debug consequences, because it may well be that objects created in the sourced script are not available in later chunks.
I am empathetic to the argument that `knitr::read_chunk()` may confuse people, though then we should recommend a "save" [version of source](https://bookdown.org/yihui/rmarkdown-cookbook/source-script.html):

````
```{r, include=FALSE}
source("your-script.R", local = knitr::knit_global())
# or sys.source("your-script.R", envir = knitr::knit_global())
```
````

**Comment by brandmaier** [comment]
I have experienced these caching issues myself and it can be very annoying. Aarons suggestion is also backed up by Yihui: https://bookdown.org/yihui/rmarkdown-cookbook/source-script.html

The other really nice and easy alternative are named chunks. I think I suggested mentioning them in an earlier PR. Here is how it's done: http://zevross.com/blog/2014/07/09/making-use-of-external-r-code-in-knitr-and-r-markdown/
